src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,570 @@
+ *     <td>Foo, Bar, &amp; Baz</td>
+ *     <td>Foo, Bar, Baz</td>
+ * <tr><th scope="row" style="text-align:left">OR</th>
+ *     <td>Foo, Bar, or Baz</td>
+ *     <td>Foo, Bar, or Baz</td>
+ *     <td>Foo, Bar, or Baz</td>
+ * <tr><th scope="row" style="text-align:left">UNIT</th>
+ *     <td>Foo, Bar, Baz</td>
+ *     <td>Foo, Bar, Baz</td>
+ *     <td>Foo Bar Baz</td>
+ * </tbody>
+ * </table>
+ * <p>
+ * Alternatively, an instance with Locale, Type, and/or Style invariant patterns
+ * can be created with {@link #getInstance(String[])}. The String array to the
+ * method specifies the delimiting patterns for the start/middle/end portion of
+ * the formatted string, as well as optional specialized patterns for two or three
+ * elements. Refer to the method description for more detail.
+ * <p>
+ * On parsing, if some ambiguity found in the input string, such as delimiting
+ * sequences are used in the input string, may produce the result that is not a
+ * round-trip with the corresponding formatting. For example, a String list
+ * ["a, b,", "c"] will be formatted as "a, b, and c", but may be parsed as
+ * ["a", "b", "c"].
+ *
+ * @implSpec This class is immutable and thread-safe
+ *
+ * @spec https://www.unicode.org/reports/tr35 Unicode Locale Data Markup Language (LDML)
+ * @since 22
+ */
+public class ListFormat extends Format {
+
+    @Serial
+    private static final long serialVersionUID = 5272525550078071946L;
+
+    private static final int START = 0;
+    private static final int MIDDLE = 1;
+    private static final int END = 2;
+    private static final int TWO = 3;
+    private static final int THREE = 4;
+    private static final int PATTERN_ARRAY_LENGTH = THREE + 1;
+
+    /**
+     * The locale to use for formatting list patterns.
+     * @serial
+     */
+    private final Locale locale;
+
+    /**
+     * The array of five pattern Strings. Each element corresponds to the Unicode LDML's
+     * `listPatternsPart` type, i.e, start/middle/end/two/three.
+     * @serial
+     */
+    private final String[] patterns;
+
+    private static final Pattern PARSE_START = Pattern.compile("(?<startBefore>.*?)\\{0}(?<startBetween>.*?)\\{1}");
+    private static final Pattern PARSE_MIDDLE = Pattern.compile("\\{0}(?<middleBetween>.*?)\\{1}");
+    private static final Pattern PARSE_END = Pattern.compile("\\{0}(?<endBetween>.*?)\\{1}(?<endAfter>.*?)");
+    private transient Pattern startPattern;
+    private transient String middleBetween;
+    private transient Pattern endPattern;
+
+    private ListFormat(Locale l, String[] patterns) {
+        locale = l;
+        this.patterns = patterns;
+        init();
+    }
+
+    private void init() {
+        // get pattern strings
+        var m = PARSE_START.matcher(patterns[START]);
+        String startBefore;
+        String startBetween;
+        if (m.matches()) {
+            startBefore = m.group("startBefore");
+            startBetween = m.group("startBetween");
```

Comment:
Do named groups perform as well as indexed groups?

Reply:
Have not measured, but I think the index would be faster. Replaced.